File: 백준/Gold/16234. 인구 이동/인구 이동.py
# 인구 이동
import sys
input = sys.stdin.readline
from collections import deque

n, L, R = map(int,input().split())
graph = []
direc = [(1, 0), (-1, 0), (0, 1), (0, -1)]

for i in range(n):
    graph.append(list(map(int,input().split())))

def bfs(unions):
    elements = []
    total = 0
    
    while unions:
        r, c = unions.popleft()

        if visited[r][c]:
            continue

        visited[r][c] = 1
        elements.append((r,c))
        cur = graph[r][c]
        total += cur
        
        for x, y in direc:
            nr = r + x
            nc = c + y

            if not (0 <= nr < n) or not (0 <= nc < n):
                continue

            if visited[nr][nc]:
                continue
            
            diff = abs(cur - graph[nr][nc])

            if L <= diff <= R:
                unions.append((nr, nc))

    population = total // len(elements)

    for r, c in elements:
        graph[r][c] = population
        
    return elements

cnt = 0
while True:
    change = False
    visited = [[0] * n for _ in range(n)]
    group = []
    for i in range(n):
        for j in range(n):
            
            if visited[i][j]:
                continue
            
            elements = bfs(deque([(i, j)]))
            
            if len(elements) > 1:
                change = True
                
    if not change:
        print(cnt)
        break
    else:
        cnt += 1

File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.io.*;
import java.util.*;

public class Main {
    static int L;
    static int R;
    static int[][] graph;
    static boolean[][] visited;
    static int[] dirX = new int[] {1, -1, 0, 0};
    static int[] dirY = new int[] {0, 0, 1, -1};

    static class Point {
        int x;
        int y;

        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static int bfs(int i, int j) {
        ArrayList<Point> elements = new ArrayList<>();
        ArrayDeque<Point> union = new ArrayDeque<>();
        union.add(new Point(i, j));
        int total = 0;

        while (!union.isEmpty()) {
            Point p = union.removeFirst();

            if (visited[p.x][p.y]) {
                continue;
            } else {
                visited[p.x][p.y] = true;
            }

            elements.add(p);
            int cur = graph[p.x][p.y];
            total += cur;

            for (int k = 0; k < 4; k++) {
                int nx = p.x + dirX[k];
                int ny = p.y + dirY[k];

                if (nx < 0 || ny < 0 || nx >= graph.length || ny >= graph.length) {
                    continue;
                } else if (visited[nx][ny]) {
                    continue;
                }

                int diff = Math.abs(cur - graph[nx][ny]);

                if (L <= diff && diff <= R) {
                    union.add(new Point(nx, ny));
                }
            }
        }

        int population = total / elements.size();

        for (Point p : elements) {
            graph[p.x][p.y] = population;
        }

        return elements.size();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        L = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        graph = new int[n][n];
        int cnt = 0;

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        while (true) {
            boolean change = false;
            visited = new boolean[n][n];
            int elements = 0;

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (visited[i][j]) {
                        continue;
                    } else {
                        elements = bfs(i, j);
                    }

                    if (elements > 1) {
                        change = true;
                    }
                }
            }
            if (!change) {
                System.out.println(cnt);
                break;
            } else {
                cnt += 1;
            }
        }
    }
}
